packages/runtime-core/src/apiSetupHelpers.ts

```diff
@@ -58,7 +58,19 @@ export function defineProps<
   PP extends ComponentObjectPropsOptions = ComponentObjectPropsOptions
 >(props: PP): Readonly<ExtractPropTypes<PP>>
 // overload 3: typed-based declaration
-export function defineProps<TypeProps>(): Readonly<TypeProps>
+// the Boolean absent props will be cast to false, so the return type
+// must be a boolean type
+export function defineProps<TypeProps>(): Readonly<
+  {
+    [K in keyof TypeProps as TypeProps[K] extends boolean | undefined
+      ? K
+      : never]-?: TypeProps[K]
```

Comment:
```suggestion
      : never]-?: NotUndefined<TypeProps[K]>
```

Reply:
> 

Without `NotUndefined`,  the '-?' will still remove 'undefined'.

Reply:
Actually, it won't when there's only `boolean | undefined` and no `?:`
```ts
interface T {
  boolAndUndefined: boolean | undefined
}

type Test = {
  [K in keyof T]-?: T[K]
}
// type Test = {
//     boolAndUndefined: boolean | undefined;
// }
```
https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgCrIN4ChnIEYD2BANgIIgAmAqpRDKBBQFz5HERwjIA+yArrXohGWAL5YsYAJ4AHFKggBnMMgC8mHMgDaAaWShkAawhSCMNAF0AtAH4WqXRbFYgA

Reply:
Thanks, modified.